src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt

```diff
@@ -72,12 +76,54 @@ class RsErrorAnnotator : Annotator, HighlightRangeExtension {
             override fun visitArrayType(o: RsArrayType) = collectDiagnostics(holder, o)
             override fun visitVariantDiscriminant(o: RsVariantDiscriminant) = collectDiagnostics(holder, o)
             override fun visitPolybound(o: RsPolybound) = checkPolybound(holder, o)
+            override fun visitTryExpr(o: RsTryExpr) = checkTryExpr(holder, o)
             override fun visitTraitRef(o: RsTraitRef) = checkTraitRef(holder, o)
         }
 
         element.accept(visitor)
     }
 
+    private fun checkTryExpr(holder: AnnotationHolder, o: RsTryExpr) {
+        val items = o.knownItems
+        val lookup = ImplLookup(o.project, items)
+        val tryTrait = items.Try ?: return
+        val fromTrait = items.From ?: return
+
+        val tryExprTy = o.expr.type
+        if (tryExprTy is TyUnknown) return
+        val errorTy = findErrorType(tryExprTy, tryTrait, lookup)
+        if (errorTy == null) {
+            RsDiagnostic.TryTraitIsNotImplemented(o, tryExprTy).addToHolder(holder)
+            return
+        }
+
+        val returnTy = findParentFnOrLambdaRetTy(o)
```

Comment:
Consider such code:
```
fn main() {
    let opt: Option<String>;
    opt.and_then(|mut s| { // note: I declared just `|mut s|`, not `|mut s| -> Option<String>`, but it infers correctly
        let tail = s.get(1..)?;
        Some(String::from(tail))
    });
}
```
In this case `o` = `s.get(1..)?` and `findParentFnOrLambdaRetTy(o)` returns `null`. Because it doesn't take into account closures with inferred type. So your `findParentFnOrLambdaRetTy` might be like this:
```
fun correctFindParentFnOrLambdaReturnType(element: RsExpr): Ty? {
    var parent = element.parent
    while (parent != null) {
        when (parent) {
            is RsFunction -> return parent.retType?.typeReference?.type
            is RsLambdaExpr -> return parent.retType?.typeReference?.type ?: (parent.type as? TyFunction)?.retType
            else -> parent = parent.parent
        }
    }
    return null
}
```

I don't know whether we should rewrite [original](https://github.com/intellij-rust/intellij-rust/blob/10f0ba3f61a58f7ac5f7c0bfa4ebe338aeb40a0e/src/main/kotlin/org/rust/ide/annotator/fixes/ConvertToTyUsingTryFromTraitFix.kt#L72) function or add a new one, because the tests for `ConvertToTyUsingTryTraitAndUnpackFix` looks like `let f = || -> Result<Bb, Ee> { let b: Bb = Bb::try_from(Aa)?; return Ok(b);};` (it's impossible to omit return type here). @vlad20012 what do you think?

Reply:
I think we should not care about it now. Will fix both quickfixes in the future